app/common/ts_util.c

```diff
@@ -0,0 +1,736 @@
+  ierr = PetscViewerBinaryRead(viewer,&ts->time_step,1,NULL,PETSC_DOUBLE);CHKERRQ(ierr);
+  ierr = PetscViewerBinaryRead(viewer,&ts->ptime_prev,1,NULL,PETSC_DOUBLE);CHKERRQ(ierr);
+  ierr = PetscViewerBinaryRead(viewer,&ts->ptime_prev_rollback,1,NULL,PETSC_DOUBLE);CHKERRQ(ierr);
+  ierr = PetscViewerBinaryRead(viewer,&ts->solvetime,1,NULL,PETSC_DOUBLE);CHKERRQ(ierr);
+  /* PETSC BUG FIX END */
+
+  ierr = TSSetType(ts, type);CHKERRQ(ierr);
+  if (ts->ops->load) {
+    ierr = (*ts->ops->load)(ts,viewer);CHKERRQ(ierr);
+  }
+
+  /* PETSC BUG FIX START */
+  if (!ts->adapt) {
+    /*
+       The reason why we create/load here is because _TSView() is gaurnteed to write out a TSAdapt object.
+       However, depending on the TSType TSAdapt may not be created by default.
+       Also, some implementations of TS will actually create / load TSAdapt _within_ ts->ops->load().
+       Hence if we create the TSAdapt (because it didn't exist), we must also load it.
+       Putting TSAdaptLoad() outside of this if statement may sometimes result in TSAdapt being attempted to be loaded twice.
+    */
+    ierr = TSAdaptCreate(PetscObjectComm((PetscObject)ts),&ts->adapt);CHKERRQ(ierr);
+    ierr = TSAdaptLoad(ts->adapt,viewer);CHKERRQ(ierr);
+  }
+  
+  /* PETSC BUG FIX END */
+  ierr = DMCreate(PetscObjectComm((PetscObject)ts),&dm);CHKERRQ(ierr);
+  ierr = DMLoad(dm,viewer);CHKERRQ(ierr);
+  ierr = TSSetDM(ts,dm);CHKERRQ(ierr);
+  if (!ts->vec_sol) {
+      ierr = DMCreateGlobalVector(ts->dm,&ts->vec_sol);CHKERRQ(ierr);
+  }
+  ierr = VecLoad(ts->vec_sol,viewer);CHKERRQ(ierr);
+  ierr = DMGetDMTS(ts->dm,&sdm);CHKERRQ(ierr);
+  ierr = DMTSLoad(sdm,viewer);CHKERRQ(ierr);
+
+  ierr = _TSAdaptMembersLoad(ts->adapt,viewer);CHKERRQ(ierr);
+
+  PetscFunctionReturn(0);
+}
+
+PetscErrorCode _TSImplLoad_RK(TS ts, PetscViewer viewer)
+{
+  PetscErrorCode ierr;
+  PetscBool      isbinary;
+  PetscInt       classid;
+  char           type[256];
+  PetscBool      same;
+
+  PetscFunctionBegin;
+  PetscValidHeaderSpecific(ts,TS_CLASSID,1);
+  PetscValidHeaderSpecific(viewer,PETSC_VIEWER_CLASSID,2);
+  ierr = PetscObjectTypeCompare((PetscObject)viewer,PETSCVIEWERBINARY,&isbinary);CHKERRQ(ierr);
+  if (!isbinary) SETERRQ(PETSC_COMM_SELF,PETSC_ERR_ARG_WRONG,"Invalid viewer; open viewer with PetscViewerBinaryOpen()");
+
+  ierr = PetscViewerBinaryRead(viewer,&classid,1,NULL,PETSC_INT);CHKERRQ(ierr);
+  if (classid != TS_FILE_CLASSID) SETERRQ(PetscObjectComm((PetscObject)ts),PETSC_ERR_ARG_WRONG,"Not TS next in file");
+  ierr = PetscViewerBinaryRead(viewer,type,256,NULL,PETSC_CHAR);CHKERRQ(ierr);
+
+  PetscObjectTypeCompare((PetscObject)ts,TSRK,&same);
+  if (same) {
+    PetscInt i,s;
+    TS_RK    *rk  = (TS_RK *)ts->data;
+    /*RKTableau tab = rk->tableau;*/
+    PetscBool FSAL;
+            
+    ierr = TSRKGetTableau(ts,&s,NULL,NULL,NULL,NULL,NULL,NULL,&FSAL);CHKERRQ(ierr);
+    for (i=0; i<s; i++) {
+      PetscBool isnest;
+
+      if (rk->YdotRHS[i]) {
```

Comment:
move VecLoad under this if?

Reply:
I know the code looks a bit stupid, but it's written this way for clarity. I want these things clearly separated.